docs/source/reference/extra-handlers.md

```diff
@@ -0,0 +1,54 @@
+# Adding extra handlers to the hub
+
+As an alternative to [Services](./services.md), it is possible to add extra tornado handlers to the hub.
+
+They can be used to show custom pages, expose additional API endpoints and add extra functionalities to the hub.
+
+## Adding the handler to the JupyterHub config
+
+Additional handlers can be created and defined in the `jupyterhub_config.py` file,
+by extending [`c.JupyterHub.extra_handlers`](../api/app.html#jupyterhub.app.JupyterHub.extra_handlers).
+
+For example:
+
+```python
+from jupyterhub.handlers.base import BaseHandler
+from tornado.web import authenticated
+
+class CustomHandler(BaseHandler):
+    @authenticated
+    def get(self):
+        self.write(self.render_template("page.html"))
+
+c.JupyterHub.extra_handlers.extend([
+    (r"custom", CustomHandler),
+])
+```
+
+Issuing a GET request to `/hub/custom` will now render the default JupyterHub page with an empty content.
+
+## Adding a new link to the default navigation bar
+
+The new handler can be added to the navigation bar by creating a new `page.html` file template with the following content:
+
+```html
+{% extends "templates/page.html" %}
+
+{% block nav_bar_left_items %}
+{{ super() }}
+<li><a href="{{base_url}}custom">Custom Handler</a></li>
+{% endblock %}
+```
+
+The `jupyterhub_config.py` file can then be updated to include the custom template:
+
+```python
+extra_template_path = "/path/to/extra/templates"
+c.JupyterHub.template_paths.insert(0, extra_template_path)
```

Comment:
```suggestion
c.JupyterHub.template_paths = [extra_template_path]
```

JupyterHub will fall back on the default templates if it doesn't find a custom template that overwrites the defaults in the `JupyterHub.template_paths` list ([docs ref](https://jupyterhub.readthedocs.io/en/stable/reference/templates.html?highlight=default%20template#custom-templates)). But overwriting the `template_paths` entirely might look more confusing than just inserting in the list. @jtpio, what do you think?

Reply:
@GeorgianaElena I pick up mixed messages between this comment and the suggestion. It seems to me like you suggest A instead of B, but the code suggestion is to make B instead of A.

I know I would be worried if I had a replace operation that I would need to provide all default templates along those I've replaced. In a situation of mine, I have updated a single template, spawn.html, but nothing else, so I just want to add a paths location for that.

Reply:
> I know I would be worried if I had a replace operation that I would need to provide all default templates along those I've replaced. In a situation of mine, I have updated a single template, spawn.html, but nothing else, so I just want to add a paths location for that.

When you do `c.JupyterHub.template_paths = [extra_template_path]`, (from my understanding), it's not a replacement of default templates: https://github.com/jupyterhub/jupyterhub/blob/a4390a1f4fd0da9f8844339920428f4c0b3649f5/jupyterhub/app.py#L2218-L2219

> @GeorgianaElena I pick up mixed messages between this comment and the suggestion. It seems to me like you suggest A instead of B, but the code suggestion is to make B instead of A.

I was just trying to let @jtpio know that there's no need to accept my code change suggestion if they consider it to be confusing. Sorry if I ended up posting a confusing sentence instead :confused:

Reply:
@GeorgianaElena ah so the config option is really be more like "extra_template_paths"?

Reply:
yeah, this is how I understand this.

Reply:
:+1: !!